### ISSUE TITLE
Support for Linux 5.8

### ISSUE THREAD

**Issue by Alexerson** [sophos/talpa#20]
Hello, 
I'm trying to compile talpa for my archlinux running the Kernel 5.8.10, but the talpa-select select command fails.

```
checking for unused task flag... ffzb_local.c: In function 'main':
ffzb_local.c:50:1: error: 'PF_KTHREAD' undeclared (first use in this function)
   50 | PF_KTHREAD | \
      | ^~~~~~~~~~
ffzb_local.c:67:27: note: in expansion of macro 'PF_TALPA_ALL'
   67 |     unsigned int result = PF_TALPA_ALL;
      |                           ^~~~~~~~~~~~
ffzb_local.c:50:1: note: each undeclared identifier is reported only once for each function it appears in
   50 | PF_KTHREAD | \
      | ^~~~~~~~~~
ffzb_local.c:67:27: note: in expansion of macro 'PF_TALPA_ALL'
   67 |     unsigned int result = PF_TALPA_ALL;
      |                           ^~~~~~~~~~~~
configure: error: compiler failure

Traceback (most recent call last):
  File "talpa_select.py", line 2035, in _action
  File "talpa_select.py", line 664, in select
  File "talpa_select.py", line 1555, in select
  File "talpa_select.py", line 1639, in build
  File "talpa_select.py", line 1769, in __try_build
  File "talpa_select.py", line 1634, in checkConfigureErrors
SelectException: exc-configure-failed

```

Any idea?

**Comment by freddyw** [comment]
I love to see support for kernel 5.8 (and soon 5.9) as well.

**Comment by Pastim** [comment]
It would be really good o have this support.

**Comment by paperclip** [comment]
Unfortunately, we are currently working on other products, so haven't been able to look at Talpa in quite a while.

**Comment by freddyw** [comment]
I have tried just about anything to make it work with kernel 5.9 and 5.10. It got it to compile but it just won't work reliably. Somehow one of my searches led me here:
https://support.sophos.com/support/s/article/KB-000034610?language=en_US

And you know what? Switching to fanotify works for me. Not sure if there are any disadvanges (just yet) but so far so good.

**Comment by Pastim** [comment]
I've given up on free Sophos for linux.  Using fanotify It started to prevent me installing normal linux updates (it took me a long while to figure out that was what was happening).  SInce support is being withdrawn for this product I figured it was time to move on.  I've paid for Eset.

**Comment by freddyw** [comment]
We came from Eset and that was not a happy marriage either. Using fanotity on Fedora 33 and 5.10.0-0.rc6 works like a treat so far ...

**Comment by Pastim** [comment]
> We came from Eset and that was not a happy marriage either. Using fanotity on Fedora 33 and 5.10.0-0.rc6 works like a treat so far ...
I'm on ubuntu 20.10, using fanotify because talpa is not supported.  On installing linux kernel updates, wine updates and others, the installer complained it was unable to find or open some files.  This baffled me for ages.  When I suspended sophos everything worked fine.  

Eset hasn't got in the way, yet....

**Comment by adiemu5** [comment]
The first issue here seems to be that a recent-ish changeto 'include/linux/sched.h' which previously looked like this:
```
#define PF_LESS_THROTTLE        0x00100000      /* Throttle me less: I clean memory */
#define PF_KTHREAD              0x00200000      /* I am a kernel thread */
```
but that now looks like this:
```
#define PF_LOCAL_THROTTLE       0x00100000      /* Throttle writes only against the bdi I write to,
                                                 * I am cleaning dirty pages from some other bdi. */
#define PF_KTHREAD              0x00200000      /* I am a kernel thread */
```

In other words, the comment for `PF_LOCAL_THROTTLE` became multi-line.   (I'm on CentOS 8, and it happened when going from 8.3 to 8.4)

This causes problems for talpa source builds because talpa's current `configure` script has the line:

```
    if ! ${EGREP} "#define PF_" $TALPA_HEADER_LOCATION/linux/sched.h > ffzb_local.c; then
```
which during a talpa source build attempt results in `ffzb_local.c` containing the lines:
```
#define PF_LOCAL_THROTTLE       0x00100000      /* Throttle writes only against the bdi I write to,
#define PF_KTHREAD              0x00200000      /* I am a kernel thread */
```
Note the lack of a closing `*/` on `PF_LOCAL_THROTTLE`'s comment, meaning that `PF_KTHREAD` is commented out, since `PF_LOCAL_THROTTLE`'s comment extends to the `*/` on `PF_KTHREAD`'s line.

Fixing the `EGREP` line in `configure` is straightforward enough using `sed` to force the lines to end with `*/`:
```
    if ! ${EGREP} "#define PF_" $TALPA_HEADER_LOCATION/linux/sched.h | ${SED} 's|..\s*$|\*/|' > ffzb_local.c; then
```
which makes `ffzb_local.c` contain:
```
#define PF_LOCAL_THROTTLE       0x00100000      /* Throttle writes only against the bdi I write t*/
#define PF_KTHREAD              0x00200000      /* I am a kernel thread */
```

From here, a manual run of `./configure` succeeds, but `make` fails with a new issue:
```
# ./configure --disable-dependency-tracking --libdir=/opt/sophos-av/talpa/current --disable-lsm
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
<snip>
config.status: config.h is unchanged
config.status: executing depfiles commands
# make
<snip>
  CC [M]  /opt/sophos-av/talpa/build/talpa-1.25.3/src/components/intercepts/vfshook_impl/vfshook_interceptor.o
/opt/sophos-av/talpa/build/talpa-1.25.3/src/components/intercepts/vfshook_impl/vfshook_interceptor.c: In function ‘processMount’:
/opt/sophos-av/talpa/build/talpa-1.25.3/src/components/intercepts/vfshook_impl/vfshook_interceptor.c:2394:55: error: ‘MS_REMOUNT’ undeclared (first use in this function); did you mean ‘MNT_UMOUNT’?
             if ( shouldinc && !(fromMount && (flags & MS_REMOUNT)) )
                                                       ^~~~~~~~~~
                                                       MNT_UMOUNT
/opt/sophos-av/talpa/build/talpa-1.25.3/src/components/intercepts/vfshook_impl/vfshook_interceptor.c:2394:55: note: each undeclared identifier is reported only once for each function it appears in
In file included from ./include/linux/kernel.h:11,
                 from /opt/sophos-av/talpa/build/talpa-1.25.3/src/components/intercepts/vfshook_impl/vfshook_interceptor.c:20:
/opt/sophos-av/talpa/build/talpa-1.25.3/src/components/intercepts/vfshook_impl/vfshook_interceptor.c: In function ‘talpaPreMount’:
/opt/sophos-av/talpa/build/talpa-1.25.3/src/components/intercepts/vfshook_impl/vfshook_interceptor.c:2493:28: error: ‘MS_BIND’ undeclared (first use in this function); did you mean ‘LAST_BIND’?
 #define VFSHOOK_MS_IGNORE (MS_BIND)
                            ^~~~~~~
./include/linux/compiler.h:77:42: note: in definition of macro ‘unlikely’
 # define unlikely(x) __builtin_expect(!!(x), 0)
                                          ^
/opt/sophos-av/talpa/build/talpa-1.25.3/src/components/intercepts/vfshook_impl/vfshook_interceptor.c:2533:31: note: in expansion of macro ‘VFSHOOK_MS_IGNORE’
     if ( unlikely( ( (flags & VFSHOOK_MS_IGNORE) != 0 ) ) )
                               ^~~~~~~~~~~~~~~~~
/opt/sophos-av/talpa/build/talpa-1.25.3/src/components/intercepts/vfshook_impl/vfshook_interceptor.c: In function ‘talpaPostMount’:
/opt/sophos-av/talpa/build/talpa-1.25.3/src/components/intercepts/vfshook_impl/vfshook_interceptor.c:2493:28: error: ‘MS_BIND’ undeclared (first use in this function); did you mean ‘LAST_BIND’?
 #define VFSHOOK_MS_IGNORE (MS_BIND)
                            ^~~~~~~
/opt/sophos-av/talpa/build/talpa-1.25.3/src/components/intercepts/vfshook_impl/vfshook_interceptor.c:2904:21: note: in expansion of macro ‘VFSHOOK_MS_IGNORE’
         if (flags & VFSHOOK_MS_IGNORE)
                     ^~~~~~~~~~~~~~~~~
make[4]: *** [scripts/Makefile.build:316: /opt/sophos-av/talpa/build/talpa-1.25.3/src/components/intercepts/vfshook_impl/vfshook_interceptor.o] Error 1
make[3]: *** [Makefile:1563: _module_/opt/sophos-av/talpa/build/talpa-1.25.3] Error 2
make[3]: Leaving directory '/usr/src/kernels/4.18.0-305.3.1.el8.x86_64'
make[2]: *** [makefile:2015: talpa_core.ko] Error 2
make[2]: Leaving directory '/opt/sophos-av/talpa/build/talpa-1.25.3'
make[1]: *** [makefile:1642: all-recursive] Error 1
make[1]: Leaving directory '/opt/sophos-av/talpa/build/talpa-1.25.3'
make: *** [makefile:724: all] Error 2
#
```
Choking on `MS_REMOUNT` and `MS_BIND`.  It seems that these define's were moved from `include/uapi/linux/fs.h` to `include/uapi/linux/mount.h`, but this was already accounted for in talpa's commit 0abbc54653a508329e5cf933708100d4ccc0d160.  Unfortunately, this assumes it's only necessary for kernel versions >= 5.0, which means it doesn't apply on CentOS 8.x where kernel features get backported into 4.18.0.

Looks like we need to modify `src/components/intercepts/vfshook_impl/vfshook_interceptor.c` from:
```
#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,0,0)
#include <uapi/linux/mount.h>
#endif
```
to:
```
#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,0,0) || defined TALPA_HAS_BACKPORTED_MOUNT
#include <uapi/linux/mount.h>
#endif
```
and then modify `configure` (well, probably actually `configure.ac`) to actually set TALPA_HAS_BACKPORTED_MOUNT in the right circumstances.

But as I've already spent all day getting this far, I'll stop here and pick this up again later.  (Since I also have the hurdle of not seeming to be able to actually get `talpa_select` to actually use my modifications, even if I package them up into `override/talpa-srcpack.tar.gz`)

**Comment by adiemu5** [comment]
Ok, so I was able to get it compiled and loading on CentOS-8.4 using talpa_select via an override source pack.

Below is the necessary diff for talpa-0.9.95 that's currently in github.  (ie, 9567685312971ff1a187e028b0a6cd02848a13ef)

To use it:
```
# Cribbed from: https://gist.github.com/paperclip/7892031
git clone https://github.com/sophos/talpa.git
cd talpa
patch -p0 --strip=1 < ../talpa-centos-8.4.diff
./configure --disable-talpa-build --enable-maintainer-mode
make talpa-srcpack.tar.gz

# Assuming you're doing this on a box that already has Sophos installed
mv talpa-srcpack.tar.gz /opt/sophos-av/talpa/override/talpa-srcpack.tar.gz
/opt/sophos-av/engine/talpa_select select
```
Here's the actual diff:
```diff
diff --git a/Makefile26.in b/Makefile26.in
index d8b7a1a..12ad50e 100644
--- a/Makefile26.in
+++ b/Makefile26.in
@@ -53,6 +53,7 @@ EXTRA_CFLAGS += -I$(src)/src -I$(src)/src/ifaces -I$(src)/include \
                @lookup_mnt_last@ \
                @get_fs_root_and_pwd@ \
                @vfsmount_namespace@ \
+               @uapimountbackport@ \
         @compile_flags@

 obj-m := talpa_linux.o talpa_core.o talpa_vcdevice.o talpa_pedevice.o talpa_pedconnector.o
diff --git a/configure.ac b/configure.ac
index aa7f9e5..74a7f2a 100644
--- a/configure.ac
+++ b/configure.ac
@@ -198,6 +198,15 @@ AC_ARG_ENABLE(backported-jiffies,
                 ],
                 [   auto_jiffiesbackport=yes])

+AC_ARG_ENABLE(backported-uapi-mount,
+                [  --enable-backported-uapi-mount      RHEL 8.4 kernel can have features backported from newer versions [[autodetected]]],
+                [   if test "$enableval" = "yes"; then
+                        AC_SUBST(uapimountbackport,"-DTALPA_HAS_BACKPORTED_UAPI_MOUNT")
+                    fi
+                    auto_uapimountbackport=no
+                ],
+                [   auto_uapimountbackport=yes])
+
 AC_ARG_ENABLE(backported-filldir,
                 [  --enable-backported-filldir    More recent filldir_t callbacks pass u64 for inode number [[autodetected]]],
                 [   if test "$enableval" = "yes"; then
@@ -878,7 +887,7 @@ else
     TALPA_LOCATE_HEADER(linux/sched.h, $kernelhdrspath)
     AC_MSG_CHECKING([for unused task flag])
     ${RM} -f ffzb_local.c ffzb_local
-    if ! ${EGREP} "#define PF_" $TALPA_HEADER_LOCATION/linux/sched.h > ffzb_local.c; then
+    if ! ${EGREP} "#define PF_" $TALPA_HEADER_LOCATION/linux/sched.h | ${SED} 's|..\s*$|\*/|' > ffzb_local.c; then
         AC_MSG_ERROR([grep error])
     fi
     echo "#define PF_TALPA_ALL \\" >> ffzb_local.c
@@ -1577,6 +1586,25 @@ if test $((kern_version_code)) -ge 132608; then
         fi
     fi

+    # Autodetect if uapi/linux/mount.h change was backported
+    # < 5.0
+    if test $((kern_version_code)) -lt 327680; then
+        if test "$auto_uapimountbackport" = "yes"; then
+            AC_MSG_CHECKING([for backported uapi mount])
+            if test "$rhel_release" = "yes"; then
+                ## 8.4
+                if test $((rhel_release_code)) -ge 2052; then
+                    AC_MSG_RESULT([detected])
+                    AC_SUBST(uapimountbackport,"-DTALPA_HAS_BACKPORTED_UAPI_MOUNT")
+                else
+                    AC_MSG_RESULT([not detected])
+                fi
+            else
+                AC_MSG_RESULT([not detected])
+            fi
+        fi
+    fi
+
     # Detect whether smbfs is present
     AC_MSG_CHECKING([for smbfs])
     TALPA_LOCATE_HEADER(linux/smb_fs.h, $kernelhdrspath, /bin/true, /bin/false , [quiet])
@@ -2118,6 +2146,7 @@ if test $((kern_version_code)) -ge 132608; then
                 s/@jiffiesbackport@/${jiffiesbackport}/; \
                 s/@filldirbackport@/${filldirbackport}/; \
                 s/@mountrcubackport@/${mountrcubackport}/; \
+                s/@uapimountbackport@/${uapimountbackport}/; \
                 s/@smbfs@/${smbfs}/; \
                 s/@buildlsm@/${lsm}/; \
                 s/@lsm265@/${lsm265}/; \
@@ -2179,6 +2208,7 @@ if test $((kern_version_code)) -ge 132608; then
                 s/@jiffiesbackport@/${jiffiesbackport}/; \
                 s/@filldirbackport@/${filldirbackport}/; \
                 s/@mountrcubackport@/${mountrcubackport}/; \
+                s/@uapimountbackport@/${uapimountbackport}/; \
                 s/@smbfs@/${smbfs}/; \
                 s/@buildlsm@/${lsm}/; \
                 s/@lsm265@/${lsm265}/; \
diff --git a/src/components/intercepts/vfshook_impl/vfshook_interceptor.c b/src/components/intercepts/vfshook_impl/vfshook_interceptor.c
index b41bdfb..343324c 100644
--- a/src/components/intercepts/vfshook_impl/vfshook_interceptor.c
+++ b/src/components/intercepts/vfshook_impl/vfshook_interceptor.c
@@ -26,7 +26,7 @@
 #include <linux/unistd.h>

 #include <linux/mount.h>
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,0,0)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,0,0) || TALPA_HAS_BACKPORTED_UAPI_MOUNT
 #include <uapi/linux/mount.h>
 #endif
 #include <linux/sched.h>
```

Now to figure out how to create a pull request...

**Comment by adiemu5** [comment]
Pull request created: https://github.com/sophos/talpa/pull/21

**Comment by paperclip** [comment]
Should be fixed by 030ed8550d58dea61e96d8c3b09e98f57ef03f11
